sys/suit/transport/worker.c

```diff
@@ -161,22 +174,73 @@ static void *_suit_worker_thread(void *arg)
     return NULL;
 }
 
-void suit_worker_trigger(const char *url, size_t len)
+/** Reap the zombie worker, or return false if it's still running */
+static bool _worker_reap(void)
 {
-    mutex_lock(&_worker_lock);
     if (_worker_pid != KERNEL_PID_UNDEF) {
         if (thread_kill_zombie(_worker_pid) != 1) {
             /* This will only happen if the SUIT thread runs on a lower
              * priority than the caller */
             LOG_WARNING("Ignoring SUIT trigger: worker is still busy.\n");
-            return;
+            return false;
         }
     }
+    return true;
+}
 
+void suit_worker_trigger(const char *url, size_t len)
+{
+    mutex_lock(&_worker_lock);
+    if (!_worker_reap()) {
+        return;
```

Comment:
I don't quit understand why this is needed - also shouldn't we then unlock the mutex again?

Reply:
That lock has always been in suit_worker_trigger before as well -- things were just split because suit_worker_trigger acquires the lock, whereas suit_worker_trigger_prepared already had it acquired in suit_worker_prepare (which, given its differing locking behavior, I've just renamed for consistency to suit_worker_try_prepare).

The worker lock is there to cover both access to the manifest buffer and to the thread. The lock is released by the thread. Therefore, the release of the lock just indicates that the thread is *almost* gone, not that it's gone for real -- that can only be made sure of by reaping the zombie state process (which is the thing we can't await for lack of thread_join functionality).

You're right on the unlocking part, that was a bug already -- it's fixed now and better documented; the complexity of it indicates that it might be a good idea to follow @kaspar030's suggestion of using a long-running thread (idempotently started on first use to retain the nice gc-sections/gc-functions properties) rather than struggling with locks around a thread stack.